python/lsst/ts/salobj/domain.py

```diff
@@ -43,6 +43,55 @@
 MAX_RANDOM_HOST = (1 << 31) - 1
 
 
+class QosSet:
+    """QoS profiles for topic, reader and writer.
+
+    Parameters
+    ----------
+    qos_path : `str` or `pathlib.Path`
+        Path to QoS XML file.
+    profile_name : `str`
+        Profile name; one of "Command", "Event", or "Telemetry".
+
+    Attributes
+    ----------
+    volatile : `bool`
+        Is this category of topics volatile?
+        Warning: this is determined from ``profile_name``
+        (because there is no way to read it from the QoS)
+        and that logic needs to be updated if the QoS file changes.
+
+    Notes
+    -----
+    The following QoS should be created elsewhere:
+    * publisher and subscriber: these depend on the DDS partition name,
+      which is only read when creating the SalInfo object.
+       QoS does not depend on the topic type,
+    * participant: this should not depend on the profile name,
+      and there's no point making 3 of them.
+    """
+
+    def __init__(self, qos_path, profile_name):
+        self.profile_name = profile_name
+        self.volatile = dict(
+            AckcmdProfile=True,
+            CommandProfile=True,
+            EventProfile=False,
+            TelemetryProfile=True,
+        )[profile_name]
+
+        self.qos_provider = dds.QosProvider(qos_path.as_uri(), profile_name)
+        self.topic_qos = self.qos_provider.get_topic_qos()
+
+        read_queue_policy = dds.HistoryQosPolicy(
+            depth=DDS_READ_QUEUE_LEN, kind=dds.DDSHistoryKind.KEEP_LAST
```

Comment:
I really hope you can get `DDS_READ_QUEUE_LEN` from the QoS file.

Reply:
Ok, I think I am missing something here. When we change the length of the read queue are we just changing the length of the python queue? I thought we should be changing the size of the DDS read queue too. If that is not the case, I don't think it makes much sense to have this feature.

Reply:
The point is to choose a value that can safely be used in the `dds.Reader.take_cond`.

Reply:
OK, here's the scoop. I cannot get the read queue length from dds, alas, but it was only really only wanted for to print warnings about the DDS read queues filling up. So I changed this to an attribute of QosSet (which allows it to have a different value for each type of message). It should still be updated to match the values in the QoS file, but if it is not, it only affects the accuracy of the warnings.

The other code that used this constant was changed as follows:
* SalInfo has a new constant `MAX_HISTORY_READ` that sets the maximum number of historical samples to read. I documented it carefully. Being smaller than the maximum length of any DDS write queue is very dangerous because it leads to historical data being mixed with new data. Being too large is no problem. So I set it to 10,000.
* The read topics now have a new constant `DEFAULT_QUEUE_LEN` which sets the default value of the `queue_len` constructor argument, which sets the length of the Python queue (not the DDS read queue). The value is 100, so the default is unchanged.

Domain no longer sets the reader QoS policy, so QoS is now *entirely* determined by the new QoS.xml file. I think this is an excellent change. I was afraid it might break something, but it appears to be fine.